Source/WebGPU/WGSL/GlobalVariableRewriter.cpp

```diff
@@ -236,10 +236,23 @@ void RewriteGlobalVariables::insertStructs()
 
             ASSERT(global.declaration->maybeTypeName());
             auto span = global.declaration->span();
+
+            auto* type = global.declaration->maybeTypeName()->resolvedType();
+            bool shouldBeReference = true;
+            if (std::get_if<Types::Texture>(type))
+                shouldBeReference = false;
+            else if (auto* primitive = std::get_if<Types::Primitive>(type)) {
+                if (primitive->kind == Types::Primitive::Sampler)
+                    shouldBeReference = false;
+            }
+
+            AST::TypeName::Ref memberType = *global.declaration->maybeTypeName();
+            if (shouldBeReference)
+                memberType = adoptRef(*new AST::ReferenceTypeName(span, WTFMove(memberType)));
```

Comment:
I think this is the right design, but how confident are we that this actually works on all metal devices that we support? The output program would be something like

```
kernel void foo(texture2d<float> foo) {
    ...
    texture2d<float> bar = foo;
    ...
}
```

I guess I should try to do the research to make sure this actually works everywhere.

Reply:
Looks like yes this is supposed to work everywhere.